rhp/v4/server.go

```diff
@@ -0,0 +1,1037 @@
+		AddV2Contract(TransactionSet, Usage) error
+		// RenewV2Contract finalizes an existing contract and adds its renewal.
+		RenewV2Contract(TransactionSet, Usage) error
+		// ReviseV2Contract atomically revises a contract and updates its sector
+		// roots and usage.
+		ReviseV2Contract(contractID types.FileContractID, revision types.V2FileContract, roots []types.Hash256, usage Usage) error
+		// ContractElement returns the contract state element for the given
+		// contract ID.
+		ContractElement(types.FileContractID) (types.ChainIndex, types.V2FileContractElement, error)
+
+		// AccountBalance returns the balance of an account.
+		AccountBalance(rhp4.Account) (types.Currency, error)
+		// CreditAccountsWithContract atomically revises a contract and credits the account.
+		CreditAccountsWithContract([]rhp4.AccountDeposit, types.FileContractID, types.V2FileContract) ([]types.Currency, error)
+		// DebitAccount debits an account.
+		DebitAccount(rhp4.Account, types.Currency) error
+	}
+
+	// Settings reports the host's current settings.
+	Settings interface {
+		RHP4Settings() rhp4.HostSettings
+	}
+
+	// A Server handles incoming RHP4 RPC.
+	Server struct {
+		hostKey                   types.PrivateKey
+		priceTableValidity        time.Duration
+		contractProofWindowBuffer uint64
+
+		chain      ChainManager
+		syncer     Syncer
+		wallet     Wallet
+		sectors    Sectors
+		contractor Contractor
+		settings   Settings
+	}
+)
+
+func (s *Server) lockContractForRevision(contractID types.FileContractID) (rev RevisionState, unlock func(), _ error) {
+	rev, unlock, err := s.contractor.LockV2Contract(contractID)
+	switch {
+	case err != nil:
+		return RevisionState{}, nil, err
+	case rev.Revision.ProofHeight <= s.chain.Tip().Height+s.contractProofWindowBuffer:
+		unlock()
+		return RevisionState{}, nil, errorBadRequest("contract too close to proof window")
+	case rev.Revision.RevisionNumber >= types.MaxRevisionNumber:
+		unlock()
+		return RevisionState{}, nil, errorBadRequest("contract is locked for revision")
+	}
+	return rev, unlock, nil
+}
+
+func (s *Server) handleRPCSettings(stream net.Conn) error {
+	settings := s.settings.RHP4Settings()
+	settings.ProtocolVersion = protocolVersion
+	settings.Prices.ValidUntil = time.Now().Add(s.priceTableValidity)
+	sigHash := settings.Prices.SigHash()
+	settings.Prices.Signature = s.hostKey.SignHash(sigHash)
+
+	return rhp4.WriteResponse(stream, &rhp4.RPCSettingsResponse{
+		Settings: settings,
+	})
+}
+
+func (s *Server) handleRPCReadSector(stream net.Conn) error {
+	var req rhp4.RPCReadSectorRequest
+	if err := rhp4.ReadRequest(stream, &req); err != nil {
+		return errorDecodingError("failed to read request: %v", err)
+	}
+
+	if err := req.Validate(s.hostKey.PublicKey()); err != nil {
+		return errorBadRequest("request invalid: %v", err)
+	}
+	prices, token := req.Prices, req.Token
+
+	if err := s.contractor.DebitAccount(token.Account, prices.RPCReadSectorCost(req.Length)); err != nil {
```

Comment:
How do you feel about a quick check whether the host has the sector before we debit the account? For two reasons:
1. A renter will be able to find out whether a host has a sector without paying
2. A renter isn't charged full price for what is essentially a 404

Especially 1. is useful or maybe essential in RHP4 since clients don't necessarily know whether we have a contract with a host.